File: packages/pdf-gen-js/src/templates/receipt.ts
import PDFDocument from 'pdfkit';
import { PDFConfig } from '../config';

export interface ReceiptData {
  storeName: string;
  storeAddress?: string;
  receiptNumber: string;
  dateTime: Date;
  items: Array<{
    name: string;
    quantity: number;
    price: number;
    total?: number;
  }>;
  subtotal: number;
  tax?: number;
  total: number;
  paymentMethod: string;
  thankYouMessage?: string;
}

export function renderReceipt(pdf: PDFDocument, config: PDFConfig, data: ReceiptData): void {
  const leftMargin = config.margins.left;
  const topMargin = config.margins.top;
  const pageWidth = pdf.page.width - config.margins.left - config.margins.right;
  const centerX = leftMargin + pageWidth / 2;

  pdf.font('Helvetica-Bold', 16).text(data.storeName, leftMargin, topMargin, { align: 'center', width: pageWidth });
  if (data.storeAddress) {
    pdf.font('Helvetica', 9).text(data.storeAddress, leftMargin, pdf.y, { align: 'center', width: pageWidth });
  }

  pdf.font('Helvetica', 9).text(`Receipt #: ${data.receiptNumber}`, leftMargin, pdf.y + 10);
  pdf.text(`Date/Time: ${data.dateTime.toLocaleString()}`);

  const itemsStartY = pdf.y + 15;
  pdf.moveTo(leftMargin, itemsStartY).lineTo(leftMargin + pageWidth, itemsStartY).stroke();

  pdf.font('Helvetica', 9);
  let yPos = itemsStartY + 10;

  data.items.forEach((item) => {
    const itemTotal = item.total || item.quantity * item.price;
    pdf.text(`${item.name}`, leftMargin, yPos);
    pdf.text(`${item.quantity}x @ $${item.price.toFixed(2)}`, leftMargin + 200, yPos, { width: 100 });
    pdf.text(`$${itemTotal.toFixed(2)}`, leftMargin + pageWidth - 60, yPos, { align: 'right' });
    yPos += 15;
  });

  pdf.moveTo(leftMargin, yPos).lineTo(leftMargin + pageWidth, yPos).stroke();

  yPos += 10;
  pdf.font('Helvetica', 9);
  pdf.text('Subtotal:', leftMargin + pageWidth - 130, yPos);
  pdf.text(`$${data.subtotal.toFixed(2)}`, leftMargin + pageWidth - 60, yPos, { align: 'right' });

  if (data.tax) {
    yPos += 15;
    pdf.text('Tax:', leftMargin + pageWidth - 130, yPos);
    pdf.text(`$${data.tax.toFixed(2)}`, leftMargin + pageWidth - 60, yPos, { align: 'right' });
  }

  yPos += 15;
  pdf.font('Helvetica-Bold', 10);
  pdf.text('Total:', leftMargin + pageWidth - 130, yPos);
  pdf.text(`$${data.total.toFixed(2)}`, leftMargin + pageWidth - 60, yPos, { align: 'right' });

  yPos += 20;
  pdf.font('Helvetica', 8).text(`Payment Method: ${data.paymentMethod}`, leftMargin, yPos);

  if (data.thankYouMessage) {
    yPos += 25;
    pdf.font('Helvetica-Bold', 10).text(data.thankYouMessage, leftMargin, yPos, { align: 'center', width: pageWidth });
  }
}

File: packages/pdf-gen-py/pdf_gen/templates/receipt.py
from typing import Any, Dict

from ..config import PDFConfig, Theme
from ..surface import Surface
from ..utils.validate import require_fields, to_date, format_currency


def render_receipt(surface: Surface, config: PDFConfig, data: Dict[str, Any], theme: Theme) -> None:
    require_fields(data, ['store_name', 'receipt_number', 'items', 'subtotal', 'total', 'payment_method'], 'receipt')

    left = config.margins['left']
    right = surface.page_width - config.margins['right']
    width = right - left
    currency = data.get('currency', 'USD')
    date_time = to_date(data.get('datetime'))

    y = config.margins['top']

    y = surface.text(left, y, data['store_name'], font='Helvetica-Bold', size=18, color=theme.primary, align='center', width=width)
    y += 2

    store_address = data.get('store_address')
    if store_address:
        y = surface.text(left, y, store_address, font='Helvetica', size=9, color=theme.muted, align='center', width=width)

    y += 12
    surface.line(left, y, right, y, color=theme.border, width=1)
    y += 12

    surface.text(left, y, f"Receipt #: {data['receipt_number']}", font='Helvetica', size=9, color=theme.text)
    y += 12
    surface.text(left, y, date_time.strftime('%b %d, %Y, %I:%M %p'), font='Helvetica', size=9, color=theme.text)
    y += 14

    surface.line(left, y, right, y, color=theme.border, width=1, dash=[2, 2])
    y += 10

    surface.text(left, y, 'ITEM', font='Helvetica-Bold', size=9, color=theme.muted, width=width * 0.5)
    surface.text(left + width * 0.5, y, 'QTY x PRICE', font='Helvetica-Bold', size=9, color=theme.muted, align='right', width=width * 0.3)
    surface.text(left + width * 0.8, y, 'TOTAL', font='Helvetica-Bold', size=9, color=theme.muted, align='right', width=width * 0.2)
    y += 14

    for item in data['items']:
        item_total = item.get('total', item['quantity'] * item['price'])

        if y + 16 > surface.page_height - config.margins['bottom']:
            surface.new_page()
            y = config.margins['top']

        surface.text(left, y, item['name'], font='Helvetica', size=10, color=theme.text, width=width * 0.5)
        surface.text(
            left + width * 0.5, y, f"{item['quantity']} x {format_currency(item['price'], currency)}",
            font='Helvetica', size=10, color=theme.text, align='right', width=width * 0.3
        )
        surface.text(left + width * 0.8, y, format_currency(item_total, currency), font='Helvetica', size=10, color=theme.text, align='right', width=width * 0.2)
        y += 16

    y += 4
    surface.line(left, y, right, y, color=theme.border)
    y += 10

    label_width = width - 90
    surface.text(left, y, 'Subtotal', font='Helvetica', size=10, color=theme.text, width=label_width)
    surface.text(left + label_width, y, format_currency(data['subtotal'], currency), font='Helvetica', size=10, color=theme.text, align='right', width=90)
    y += 16

    tax = data.get('tax')
    if tax:
        surface.text(left, y, 'Tax', font='Helvetica', size=10, color=theme.text, width=label_width)
        surface.text(left + label_width, y, format_currency(tax, currency), font='Helvetica', size=10, color=theme.text, align='right', width=90)
        y += 16

    surface.text(left, y, 'TOTAL', font='Helvetica-Bold', size=13, color=theme.primary, width=label_width)
    surface.text(left + label_width, y, format_currency(data['total'], currency), font='Helvetica-Bold', size=13, color=theme.primary, align='right', width=90)
    y += 24

    y = surface.text(left, y, f"Payment method: {data['payment_method']}", font='Helvetica', size=9, color=theme.muted, align='center', width=width)
    y += 14

    thank_you = data.get('thank_you_message')
    if thank_you:
        surface.text(left, y, thank_you, font='Helvetica-BoldOblique', size=11, color=theme.primary, align='center', width=width)
